### PR TITLE
[intrin]support fmod for cuda

### PR DESCRIPTION
**Pull request by xqdan**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

@siju-samuel @PariksheetPinjari909 @Huyuwei please review, thanks!

### CODE REVIEW

File: python/tvm/intrin.py [hunk 1]
```diff
@@ -376,6 +376,22 @@ def popcount(x):
     """
     return call_pure_intrin(x.dtype, "popcount", x)
 
+def mod(x, y):
```
Comment (tqchen) [inline python/tvm/intrin.py]: rename to fmod as we already have Mod in the ir node

File: src/lang/ir_operator.cc [hunk 2]
```diff
@@ -450,4 +450,10 @@ Expr prod(Expr source, Array<IterVar> rdom) {
   return ir::Reduce::make(combiner, {source}, rdom, make_const(Bool(1), true), 0);
 }
 
+Expr mod(Expr x, Expr y) {
```
Comment (tqchen) [inline src/lang/ir_operator.cc]: rename to fmod

File: tests/python/integration/test_ewise.py [hunk 3]
```diff
@@ -38,6 +38,43 @@ def check_device(device, host="stackvm"):
     check_device("cuda", "llvm")
     check_device("vulkan")
 
+def test_mod():
```
Comment (tqchen) [inline tests/python/integration/test_ewise.py]: test_fmod

File: tests/python/integration/test_ewise.py [hunk 4]
```diff
@@ -245,3 +282,4 @@ def check_device(device):
     test_add()
     test_log_pow_llvm()
     test_popcount()
+    test_mod()
```
Comment (siju-samuel) [inline tests/python/integration/test_ewise.py]: test_mod-> test_fmod

File: src/lang/ir_operator.cc [hunk 5]
```diff
@@ -450,4 +450,10 @@ Expr prod(Expr source, Array<IterVar> rdom) {
   return ir::Reduce::make(combiner, {source}, rdom, make_const(Bool(1), true), 0);
 }
 
+Expr fmod(Expr x, Expr y) {
+  BinaryOpMatchTypes(x, y);
+  CHECK(x.type().is_float()) << "mod only applies to float";
```
Comment (siju-samuel) [inline src/lang/ir_operator.cc]: fmod

File: python/tvm/intrin.py [hunk 6]
```diff
@@ -376,6 +376,22 @@ def popcount(x):
     """
     return call_pure_intrin(x.dtype, "popcount", x)
 
+def fmod(x, y):
+    """Take mod cast of input x and y
```
Comment (siju-samuel) [inline python/tvm/intrin.py]: Suggest to  make the docstring slightly detailed.

**Comment by tqchen** [comment]
please also check other backends(vulkan, opencl, metal, rocm, nvptx) to add this function, you might need to lookup the specific reference of each backend

**Comment by xqdan** [comment]
openCL: https://manpages.debian.org/testing/opencl-1.2-man-doc/fmod.3clc.en.html
metal: https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf
rocm: https://rocm-documentation.readthedocs.io/en/latest/ROCm_Compiler_SDK/ocml.html?highlight=fmod

**Comment by tqchen** [comment]
Vulkan's intrinsics are defined here https://github.com/dmlc/tvm/blob/master/src/codegen/spirv/intrin_rule_spirv.cc#L33

**Comment by xqdan** [comment]
@tqchen thanks, I mean the link of vulkan reference doc

**Comment by tqchen** [comment]
These macros are defined in ```#include <GLSL.std.450.h>``` (they are not part of Vulkan doc, but more of GLSL standard)

**Comment by kun-zh** [comment]
@tqchen thanks
Vulkan, no fmod https://www.khronos.org/registry/spir-v/specs/1.0/GLSL.std.450.html

**Comment by xqdan** [comment]
@tqchen thanks
Vulkan: no fmod https://www.khronos.org/registry/spir-v/specs/1.0/GLSL.std.450.html

We can open a thread on discussion to record the reference links so that others can get them easily

**Comment by tqchen** [comment]
Thanks @xqdan @siju-samuel this is now merged. I agree that having a documented page in the docs/dev might be helpful for developers

**Pull request merged**